dashboard/test/models/lesson_test.rb

```diff
@@ -140,6 +140,99 @@ class StageTest < ActiveSupport::TestCase
     assert_equal '/', lesson2.next_level_path_for_lesson_extras(@student)
   end
 
+  test 'raise error if lesson with no levels' do
+    script = create :script
+    lesson_group = create :lesson_group, script: script
+
+    raw_lessons = [
+      {
+        key: "Lesson1",
+        name: "Lesson 1",
+        script_levels: []
+      }
+    ]
+
+    counters = LessonGroup::Counters.new(0, 0, 0, 0)
+
+    raise = assert_raises do
+      Lesson.add_lessons(script, lesson_group, raw_lessons, counters, nil, nil)
+    end
+    assert_equal 'Lessons must have at least one level in them.  Lesson: Lesson 1.', raise.message
+  end
+
+  test 'raises error when creating invalid lockable lessons' do
+    script = create :script
+    lesson_group = create :lesson_group, script: script
+    create :level, name: 'Level1'
+    create :level, name: 'LockableAssessment1'
+
+    raw_lessons = [
+      {
+        key: "Lesson1",
+        name: "Lesson 1",
+        lockable: true,
+        script_levels: [
+          {levels: [{name: "LockableAssessment1"}], assessment: true},
+          {levels: [{name: "Level1"}]}
+        ]
+      }
+    ]
+    counters = LessonGroup::Counters.new(0, 0, 0, 0)
+
+    raise = assert_raises do
+      Lesson.add_lessons(script, lesson_group, raw_lessons, counters, nil, nil)
+    end
+    assert_equal 'Expect lockable lessons to have an assessment as their last level. Lesson: Lesson 1', raise.message
+  end
+
+  test 'creates lessons correctly' do
+    script = create :script
+    lesson_group = create :lesson_group, script: script
+    create :level, name: 'Level1'
+    create :level, name: 'Level2'
+    create :level, name: 'Level3'
+    create :level, name: 'Level4'
+
+    raw_lessons = [
+      {
+        key: "L1",
+        name: "Lesson 1",
+        script_levels: [
+          {levels: [{name: "Level1"}]},
+          {levels: [{name: "Level2"}]}
+        ]
+      },
+      {
+        key: "L2",
+        name: "Lesson 2",
+        script_levels: [
+          {levels: [{name: "Level3"}]}
+        ]
+      },
+      {
+        key: "L3",
+        name: "Lesson 3",
+        lockable: true,
+        script_levels: [
+          {levels: [{name: "Level3"}], assessment: true}
+        ]
+      }
+    ]
+    counters = LessonGroup::Counters.new(0, 0, 0, 0)
```

Comment:
Since this test case succeeds, should we also assert that counters is updated correctly?

Reply:
Done!